Std/Lean/Meta/LazyDiscrTree.lean

```diff
@@ -0,0 +1,822 @@
+import Lean.Meta.DiscrTree
+import Std.Data.Nat.Init.Lemmas
+
+namespace Lean.Meta
+
+namespace LazyDiscrTree
+
+namespace MatchClone
+open DiscrTree
+
+private def tmpMVarId : MVarId := { name := `_discr_tree_tmp }
```

Comment:
Could we have a summary of which parts of this are copy-pasted from Lean and which parts diverge?

I am tempted to suggest that this is a valid use case for `open private`!

Reply:
The `MatchClone` namespace only contains copied definitions.  I've added a comment.

I'm trying to minimize forward dependencies if discrimination trees for Simp change so I've also copied over Key.